source/extensions/transport_sockets/alts/config.cc

```diff
@@ -68,7 +68,9 @@ class AltsSharedState : public Singleton::Instance {
   ~AltsSharedState() override { grpc_alts_shared_resource_dedicated_shutdown(); }
 
 private:
+#ifdef ENVOY_GOOGLE_GRPC
```

Comment:
Why do we initialize this in this one case vs. allowing the main one to do it?

Reply:
Prior to #8293 we were relying on ProcessWide to initialize google-grpc, but that was harmful to deterministic memory behavior because grpc recently made a change where they started running some allocations in a thread after grpc_init().

So that change removed grpc_init from ProcessWide, so we wouldn't get it during all tests. And it put it explicitly in MainCommon so we *would* get it for all prod, making that a prod-neutral change.

That shifts the burden of initializing grpc to tests to classes that need grpc.

Reply:
Right makes sense, but if we are initializing it in main, why do we need it explicitly here in alts config? Should we just initialize it in the ALTS test?

Reply:
Sure we can initialize in the test instead of the class.

The instantiation-counter makes initialization idempotent but maybe it's cleaner to take the init out of the classes and assert instead that we are only initializing once.

We'd just have to make sure we covered all the tests that might depend on the class. This might not be too bad. We could also have GoogleGrpcContext provide a predicate so that the class could assert that grpc has been initialized.

Happy to iterate on that to make this cleaner.

Reply:
I'm mostly just trying to understand the reason why it is the way it is. I'm fine with whatever we decide given sufficient comments. @htuch @lizan any opinion here?

Reply:
I think having `GoogleGrpcContext` where Google gRPC is directly used is fine, and that's the idea of having GoogleGrpcContext class, so let's make a clear comment there.

Reply:
+1, fine here with comment.

Reply:
added comment here.